plugins/README.adoc

```diff
@@ -7,3 +7,256 @@ NOTE: Work in Progress...
+
+----
+
+====
+
+== A complex plugin argspec should be derived from its documentation runtime if the plugin’s source is manually maintained.
+[%collapsible]
+====
+Explanations::
+A complex plugin argspec should be derived from its documentation runtime if the plugin’s source is manually maintained.  ( >30 keys or depth>2) (action plugins and control node execution modules)
+
+Rationale::
+Hand-crafting large and complex argument specifications can be error-prone and difficult to maintain.  Programmatically generating the spec from DOCUMENTATION (already a structured data object) ensures consistency and accuracy. 
+
+====
+
+== Formatting of manually maintained plugin argspecs
+[%collapsible]
+====
+Explanations::
+Ensure a consistent approach to the way commplex argument_specs are formatted within a collection.
+
+Rationale::
+When hand-writing a complex argspec, the author may choose to build up to data structure from multiple dictionaries or vars.
+Other authors may choose to implement a complex, nested argspec as a single dictionary.
+Within a single collection, select one style and use it consistently.
+
+Examples::
+Single dictionary:
+https://github.com/ansible-collections/cisco.nxos/blob/3.0.0/plugins/module_utils/network/nxos/argspec/bgp_global/bgp_global.py
+
+Multiple dictionaries:
+https://github.com/ansible-collections/community.aws/blob/stable-3/plugins/modules/ec2_scaling_policy.py#L355-L370
+https://github.com/ansible-collections/amazon.cloud/blob/0.1.0/plugins/modules/backup_report_plan.py#L182-L234
+====
+
+
+== A plugin’s entry file should not contain more than 500 lines
+[%collapsible]
+====
+Explanations::
+A plugin’s entry file should not contain more than 500 lines
+
+Rationale::
+Long and complex code files can be difficult to maintain.
+Wherever possible, reusable functions and classes, such as those for data validation or manipulation, should be moved to module_utility or plugin_utility files and imported into plugins.
+This keeps the plugin code easier to read and maintain.
+500 lines is a general guideline that should serve most cases.
+====
+
+== Action plugins should be used when the code runs on the control node.
+[%collapsible]
+====
+Explanations::
+Use an action plugin for code that needs to run on the controller, where the host would be `localhost`.
+Rationale::
+Action plugins let you integrate local processing and local data with module functionality.
+Action plugins act in conjunction with modules to execute the actions required by playbook tasks.
+Using an action plugin will cause Ansible to skip creating a temporary tarball of execution code and runs the code directly on the controller host.
+
+Examples::
+See the `ansible.snmp.get` https://github.com/ansible-network/ansible.snmp/blob/main/plugins/action/get.py[action plugin] and associated https://github.com/ansible-network/ansible.snmp/blob/main/plugins/modules/get.py[module].
+
+====
+
+== Design decisions accounting for “jump hosts” should be avoided.
```

Comment:
I was struggling with writers block on this one.  If anyone would like to see different verbiage here I welcome suggestions.

Reply:
I'm not sure to understand what you're trying to express. You seem to be hinting at a specific bad decision that you'd like to avoid in the future, and I'm probably missing this context. A lot of modules use environment variables, so I don't know why it should be a bad thing overall.

Reply:
@cidrblock I believe the jump host recommendation may have been yours.  How would you elaborate on this one?